pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		op, err = a.globalForwardingRulesClient.Insert(ctx, req)
+		if err != nil {
+			return fmt.Errorf("creating ComputeForwardingRule %s failed: %w", a.resourceID, err)
+		}
+		log.V(2).Info("successfully created ComputeForwardingRule", "name", a.resourceID)
+		// Get the created resource
+		created, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return err
+			}
+			return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+		}
+	} else {
+		req := &computepb.InsertForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Region:                 a.Location,
+			Project:                a.projectID,
+		}
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		op, err = a.forwardingRulesClient.Insert(ctx, req)
+		if err != nil {
+			return fmt.Errorf("creating ComputeForwardingRule %s failed: %w", a.resourceID, err)
+		}
+		log.V(2).Info("successfully created ComputeForwardingRule", "name", a.resourceID)
+		// Get the created resource
+		created, err = a.forwardingRulesClient.Get(ctx, getReq)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return err
+			}
+			return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+		}
+	}
+
+	status := &krm.ComputeForwardingRuleStatus{
+		CreationTimestamp: op.Proto().CreationTimestamp,
+		SelfLink:          created.SelfLink,
+	}
+	return setStatus(u, status)
+}
+
+func (a *forwardingRuleAdapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating ComputeForwardingRule", "name", a.resourceID)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	forwardingRule := ComputeForwardingRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	forwardingRule.Name = direct.LazyPtr(a.resourceID)
+	forwardingRule.Labels = desired.Labels
+
+	// Patch only support update on networkTier field, which KCC does not support yet.
+	// Use setTarget and setLabels to update target and labels fields.
+	var err error
+	// TODO(yuhou): Checked the realGCP logs, setLabels request is being sent even when there are no updates to labels.
+	// That might because of the generated labelsFingerPrint?
+	if a.Location == "global" {
+		setLabelsReq := &computepb.SetLabelsGlobalForwardingRuleRequest{
+			Resource:                       a.resourceID,
+			GlobalSetLabelsRequestResource: &computepb.GlobalSetLabelsRequest{LabelFingerprint: a.actual.LabelFingerprint, Labels: forwardingRule.Labels},
+			Project:                        a.projectID,
+		}
+		_, err = a.globalForwardingRulesClient.SetLabels(ctx, setLabelsReq)
```

Comment:
What does SetLabels return?  The worry is always that it returns an operation and that we need to wait for that operation to complete

Reply:
it returns an operation. I add a check for op.Wait()